### PR TITLE
gcc10 fixes and code cleanup

### LINKED ISSUE #30: Replace `#include <mpif.h>` with `use mpi` on default platforms
Some hiccups on our system's wrapper scripts revealed that FMS models are still doing `#include <mpif.h>` to import the MPI symbols, even though the standard recommends the module (`use mpi`).

Currently only the `sgi_mipspro` flag enables `use mpi`.  The default is to use the CPP `#include` statement.

----

To be annoyingly pedantic (sorry!), here are some quotes from the MPI 3.1 standard.

p605:

> **INCLUDE ’mpif.h’:** This method is described in Section 17.1.4.  The use of the include file `mpif.h` is strongly discouraged starting with MPI-3.0, because this method neither guarantees compile-time argument checking nor provides sufficient techniques to solve the optimization problems with nonblocking calls, and is therefore inconsistent with the Fortran standard.  It exists only for backwards compatibility with legacy MPI applications.

p611:

> The use of the `mpif.h` include file is strongly discouraged and may be deprecated in a future
version of MPI.

The comments above refer to the Fortran include statement (`include 'mpif.h'`); the standard doesn't even consider using the CPP include, which is probably even more strongly discouraged.

----

Rather than send a patch, I figured that you guys might be more aware of any odd platforms that still require this and would be in a better position to make the change.

### PR DESCRIPTION
**Pull request by bensonr**
**Description**
Deprecated the use of netcdf.inc and mpif.h in favor of provided F90 fortran modules in FMS.  Fixed bindings in mpp_io, horiz_interp, and drifters related to interface handling.
Performed general outdated macro cleanup
Remove FFIO (assign) logic
Remove SHMEM and PSET bindings in mpp layer
Removed threadloc.c and nsclock.c as they were no longer used.
Updated autotools for the removal of files and removed fortran tests for mpif.h and netcdf.inc in configure.ac
Updated CMakelists.txt for the removal of files

Fixes #30 #306 #561

**How Has This Been Tested?**
Autotools build (gcc10 and Intel 2020 update2) and unit tests run.  Ran cases for ESM4, CM4, SPEAR, and X-SHiELD.  Also ran one SHiELD RTS C48 case.

**Checklist:**
- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] Any dependent changes have been merged and published in downstream modules
- [X] New check tests, if applicable, are included
- [X] `make distcheck` passes

### CODE REVIEW

**Review by thomas-robinson — CHANGES_REQUESTED** [review]

**Review by marshallward — APPROVED** [review]

**Review by wrongkindofdoctor — APPROVED** [review]

File: mpp/include/mpp_io_connect.inc [hunk 1]
```diff
@@ -179,7 +150,7 @@
 !                                                                            !
 !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
     subroutine mpp_open( unit, file, action, form, access, threading, &
-                         fileset, iospec, nohdrs, recl,               &
```
[the remark this answers is not in the document]
Comment (bensonr) [reply mpp/include/mpp_io_connect.inc:182]: @thomas-robinson I understand your concern, but I did searches of the codebases for currently supported models prior to removal.  I could not find any references to iospec outside of FMS  I do know this option hasn't been viable for at least 20 years and the presence of the optional argument is a no-op.  I don't think we have any danger in removing it.

[the remark this answers is not in the document]
Comment (thomas-robinson) [reply mpp/include/mpp_io_connect.inc:182]: OK then, I will resolve this and approve the changes.

**Review by thomas-robinson — DISMISSED** [review]
Please remember to mention that `iospec_ieee32` was removed from the namelist in the release notes

**Review by wrongkindofdoctor — APPROVED** [review]

File: mpp/include/mpp_write_2Ddecomp.h [hunk 2]
```diff
@@ -130,20 +130,16 @@
 !write time information if new time
           if( newtime )then
               if( KIND(time).EQ.DOUBLE_KIND )then
-                  error = NF_PUT_VAR1_DOUBLE( mpp_file(unit)%ncid, mpp_file(unit)%id, mpp_file(unit)%time_level, time )
+                  error = NF_PUT_VAR1_DOUBLE( mpp_file(unit)%ncid, mpp_file(unit)%id, mpp_file(unit:unit)%time_level, time )
```
Comment (rem1776) [inline mpp/include/mpp_write_2Ddecomp.h:133]: Why was this call changed to use `mpp_file(unit:unit)%time_level` as opposed to just `mpp_file(unit)%time_level` ?

**Review by rem1776 — APPROVED** [review]

File: mpp/include/mpp_write_2Ddecomp.h (hunk 2 above)
Comment (bensonr) [reply mpp/include/mpp_write_2Ddecomp.h:133]: The interface checking in F90 throws an error about array/scalar mismatch.  The compiler interprets mpp_file(unit) as the array mpp_file(unit:).  By restricting the length of mpp_file with (unit:unit) to a single entry, the interface checking now sees mpp_file(unit:unit)%time_level properly.

**Review by thomas-robinson — APPROVED** [review]

**Pull request merged**